csi/controller.go

```diff
@@ -1523,13 +1524,42 @@ func waitForVolumeId(ctx context.Context, req *csi.CreateVolumeRequest, volumes
 		}
 		return "", true, fmt.Errorf("volume id for %s is not generated", req.GetName())
 	}, 120*time.Second, 10*time.Second)
+
 	if err != nil {
 		return "", err
 	}
 
 	return id.(string), err
 }
 
+func waitForLocalSnapshot(ctx context.Context, csiSnapshotID string, cloudBackupClient api.OpenStorageCloudBackupClient, logger *logrus.Entry) (*api.SdkCloudBackupStatus, error) {
+	// the create snapshot operation does not take local snapshot immediately. As per CSI spec, the CreationTime of the snapshot
+	// is wen the local snapshot was taken, hence wait till local snap is available
+	snapStatus, err := task.DoRetryWithTimeout(func() (interface{}, bool, error) {
+		backupStatusResponse, err := cloudBackupClient.Status(ctx, &api.SdkCloudBackupStatusRequest{
+			TaskId: csiSnapshotID,
+		})
+		if err != nil {
+			logger.Errorf("Failed to get snapshot status after snapshot create")
+			return nil, true, err
+		}
+		snapshotStatus, ok := backupStatusResponse.Statuses[csiSnapshotID]
+		if !ok {
+			return nil, true, status.Errorf(codes.NotFound, "Snapshot %s not found in status map", csiSnapshotID)
+		}
```

Comment:
This looks good to me. Only suggestion, is can we really test it by adding a sleep in porx code and see how this behaves when porx code doesn't respond within 2 minutes.

Reply:
Tested this out, byt commenting out my fix to send the local snap time. Initially the snapshot did not happen and the following error came 
```Status:
  Bound Volume Snapshot Content Name:  snapcontent-ae7cc560-b837-480d-bd27-acf1a11232ba
  Error:
    Message:     Failed to check and update snapshot content: failed to take snapshot of the volume 582374246902790351: "rpc error: code = Unknown desc = local snapshot for the cloud snapshot is not ready"
    Time:        2024-10-29T04:46:21Z
```
Then I restarted px with the fix, and the snaphsot completed
```
Events:
  Type    Reason            Age   From                 Message
  ----    ------            ----  ----                 -------
  Normal  CreatingSnapshot  30m   snapshot-controller  Waiting for a snapshot default/new-snapshot-test-10 to be created by the CSI driver.
  Normal  SnapshotCreated   16m   snapshot-controller  Snapshot default/new-snapshot-test-10 was successfully created by the CSI driver.
  Normal  SnapshotReady     16m   snapshot-controller  Snapshot default/new-snapshot-test-10 is ready to use.
```